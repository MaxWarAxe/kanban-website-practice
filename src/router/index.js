import { createRouter, createWebHistory } from 'vue-router'

const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes: [
    { path: '/tasks/:id', component: () => import('@/components/TaskEdit.vue')}
  ]
})

export default router
