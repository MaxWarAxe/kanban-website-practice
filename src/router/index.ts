import { createRouter, createWebHistory } from 'vue-router'

const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes: [
    { path: '/tasks/:id', component: () => import('@/components/TaskEdit.vue')},
    { path: '/users/add', component: () => import('@/components/PerformerEdit.vue')}
  ]
})

export default router
