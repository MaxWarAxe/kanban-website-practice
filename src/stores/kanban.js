import { defineStore } from "pinia"
import { ref } from 'vue'
import performersJSON from '../../public/performers.json'
import tasksJSON from '../../public/tasks.json'
import columnsJSON from '../../public/columns.json'

export const useKanbanStore = defineStore('kanban', () => { 
    let columnId = ref(3)
    let taskId = ref(3)
    let performerId = ref(3)
    let columns = ref([])
    let tasks = ref([])
    let performers = ref([])
    let dragging = ref(false)

    function init(){
        // performers.value = performersJSON
        // tasks.value = tasksJSON.map((task)=>{task.performer = performers.value.find(performer => performer.id == task.performerId); return task})
        // columns.value = columnsJSON.forEach((column)=>)
        performers.value = [{id : 1, name : 'Иван Иванчевич'},
                            {id : 2, name : 'Григорий Григориевич'},
                            {id : 3, name : 'Максим Максомов'}]

        tasks.value = [{id : 1, title : 'Задача по проге', content : '1. Сделай то 2. Сделай сё 3. Сделай это', performer: performers.value[0]},
                       {id : 2, title : 'Задача по тестированию', content : 'Напиши тесты по тому, по сему', performer : performers.value[1]},
                       {id : 3, title : 'Вводим в эксплуатацию', content : 'Текст какой-то Текст какой-то Текст какой-то Текст какой-то', performer : performers.value[2]}]
        
        
        columns.value = [{id: 1, header : 'Todo', items : [tasks.value[0]]},
                        {id : 2, header : 'Tests', items : [tasks.value[1]]},
                        {id : 3, header : 'Done', items : [tasks.value[2]]}]
                           console.log(JSON.stringify(columns.value))      
    }

    function removeItemFromColumn(itemID, columnID){
    var column = columns.value.find(column => column.id == columnID)
        column.items = column.items.filter(function(item) {
            return item.id != itemID;
        });
    }

    function addItemToColumn(itemID, columnID){
        var column = columns.value.find(column => column.id == columnID)
        column.items.push(tasks.value.find(task => task.id == itemID))
    }

    function moveItem(itemID, previousColumnID, newColumnID){
        removeItemFromColumn(itemID, previousColumnID)
        addItemToColumn(itemID, newColumnID)
    }

    function addNewColumn(){
        columns.value.push({id : ++columnId.value, header : 'Новая колонка', items : []})
    }

    function removeColumn(columnID){
        var column = columns.value.find(column => column.id == columnID)
        columns.value = columns.value.filter(function(column) {
            return column.id != columnID;
        });
    }

    function addNewItem(columnID){
        var column = columns.value.find(column => column.id == columnID)
        var newTask = {id : ++taskId.value, title : 'Новая задача', content : 'Новый текст', performer : null}
        tasks.value.push(newTask)
        column.items.push(newTask)
    }

    function startDrag(event,itemID,currentColumnID){
        dragging.value = true
        event.dataTransfer.dropEffect = 'move'
        event.dataTransfer.effectAllowed = 'move'
        event.dataTransfer.setData('itemID', itemID)
        event.dataTransfer.setData('columnID',currentColumnID)
    }

    function onDropColumn(event, columnID) {
        const itemID = event.dataTransfer.getData('itemID')
        const previousColumnID = event.dataTransfer.getData('columnID')
        moveItem(itemID,previousColumnID, columnID)
    }    

    function dragEnd(){
        dragging.value = false
    }

    function removeTask(taskID, columnID){
        removeItemFromColumn(taskID, columnID)
        tasks.value = tasks.value.filter(function(task){
            return task.id != taskID;
        })
    }

    function onRemoveDrop(event){
        const itemID = event.dataTransfer.getData('itemID')
        const previousColumnID = event.dataTransfer.getData('columnID')
        removeTask(itemID,previousColumnID)
    }

    function getTask(taskID){
        return tasks.value.find(task => task.id == taskID)
    }

    function addNewPerformer(name){
        performers.value.push({id : ++performerId.value, name : name})
    }

    function removePerformer(performerID){
        tasks.value.forEach(function(task){
            if(task.performer != null && task.performer.id == performerID)
                task.performer = undefined
        })
        performers.value = performers.value.filter(function(performer){
            return performer.id != performerID;
        })
    }

    return{
        onRemoveDrop,
        dragging,
        dragEnd,
        columns,
        tasks,
        performers,
        init,
        moveItem,
        addNewColumn,
        removeColumn,
        addNewItem, 
        startDrag,
        onDropColumn,
        getTask,
        addNewPerformer,
        removePerformer
    }
})