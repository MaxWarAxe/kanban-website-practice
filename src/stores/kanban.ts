import { defineStore } from "pinia"
import { Ref, ref } from 'vue'
import columnsJSON from '@/assets/data/columns.json' 
import type { Column } from '@/classes/Column.ts'
import type { Task } from '@/classes/Task.ts'
import type { Performer } from '@/classes/Performer.ts'

export const useKanbanStore = defineStore('kanban', () => { 
    const columnId : Ref<number> = ref(1)
    const taskId : Ref<number> = ref(1)
    const performerId : Ref<number> = ref(1)
    const columns : Ref<Column[]> = ref(new Array<Column>)
    const tasks : Ref<Task[]> = ref(new Array<Task>)
    const performers : Ref<Performer[]> = ref(new Array<Performer>)
    const dragging : Ref<boolean> = ref(false)

    function init(){   
        columns.value = columnsJSON
        for (let i = 0; i < columns.value.length; i++) {
            for (let j = 0; j < columns.value[i].items.length; j++){
                tasks.value.push(columns.value[i].items[j])
            }
        }
        for (let i = 0; i < tasks.value.length; i++) {
            if (tasks.value[i].performer == null)
                continue
            if (!performers.value.find((performer : Performer)=>performer.id == tasks.value[i].performer.id)) {
                performers.value.push(tasks.value[i].performer)
            }   
        }
        columnId.value = Math.max(...columns.value.map((column : Column) => column.id))
        taskId.value = Math.max(...tasks.value.map((task : Task) => task.id))
        performerId.value = Math.max(...performers.value.map((performer : Performer) => performer.id))
    }

    function removeItemFromColumn(itemID : number, columnID : number) {
        const column = columns.value.find((column : Column) => column.id === columnID)
        column.items = column.items.filter((item : Task) => item.id !== itemID)
    }

    function addItemToColumn(itemID : number, columnID : number) {
        const column = columns.value.find((column : Column) => column.id == columnID)
        column.items.push(tasks.value.find((task : Task) => task.id == itemID))
    }

    function moveItem(itemID : number, previousColumnID : number, newColumnID : number) {
        removeItemFromColumn(itemID, previousColumnID)
        addItemToColumn(itemID, newColumnID)
    }

    function addNewColumn() {
        const newColumnID = columnId.value++
        columns.value.push({id : newColumnID, header : 'Новая колонка', items : []})
    }

    function removeColumn(columnID : number) {
        columns.value = columns.value.filter((column : Column) => column.id != columnID)
    }

    function addNewItem(columnID : number) {
        const column = columns.value.find((column : Column) => column.id == columnID)
        const newTaskId = taskId.value++
        const newTask = {id : newTaskId, title : 'Новая задача', content : 'Новый текст', performer : null}
        tasks.value.push(newTask)
        column.items.push(newTask)
    }

    function startDrag(event : DragEvent, itemID : number,currentColumnID : number) {
        dragging.value = true
        event.dataTransfer.dropEffect = 'move'
        event.dataTransfer.effectAllowed = 'move'
        event.dataTransfer.setData('itemID', String(itemID))
        event.dataTransfer.setData('columnID',String(currentColumnID))
    }

    function onDropColumn(event : DragEvent, columnID : number) {
        const itemID = Number(event.dataTransfer.getData('itemID'))
        const previousColumnID = Number(event.dataTransfer.getData('columnID'))
        moveItem(itemID,previousColumnID, columnID)
    }    

    function dragEnd(){
        dragging.value = false
    }

    function removeTask(taskID : number, columnID : number) {
        removeItemFromColumn(taskID, columnID)
        tasks.value = tasks.value.filter(task => task.id != taskID)
    }

    function onRemoveDrop(event : DragEvent) {
        const itemID = Number(event.dataTransfer.getData('itemID'))
        const previousColumnID = Number(event.dataTransfer.getData('columnID'))
        removeTask(itemID,previousColumnID)
    }

    function getTask(taskID : number) {
        return tasks.value.find(task => task.id == taskID)
    }

    function addNewPerformer(name : string) {
        const newPerformerId = performerId.value++
        performers.value.push({id : newPerformerId, name : name})
    }

    function removePerformer(performerID : number) {
        tasks.value.forEach(function(task){
            if (task.performer != null && task.performer.id == performerID)
                task.performer = undefined
        })
        performers.value = performers.value.filter(performer=>performer.id != performerID)
    }

    return {
        onRemoveDrop,
        dragging,
        dragEnd,
        columns,
        tasks,
        performers,
        init,
        moveItem,
        addNewColumn,
        removeColumn,
        addNewItem, 
        startDrag,
        onDropColumn,
        getTask,
        addNewPerformer,
        removePerformer
    }
})