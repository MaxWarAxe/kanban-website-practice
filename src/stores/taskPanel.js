import { defineStore } from "pinia"
import { ref } from 'vue'

export const useTaskPanelStore = defineStore('taskPanel', () => { 
    let visible = ref(false)
    
    function show(){
        visible.value = true
    }

    function close(){
        visible.value = false
    }

    return {visible,show,close}
})