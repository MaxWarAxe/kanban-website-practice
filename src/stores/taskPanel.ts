import { defineStore } from "pinia"
import { Ref, ref } from 'vue'

export const useTaskPanelStore = defineStore('taskPanel', () => { 
    const visible : Ref<boolean> = ref(false)
    
    function show(){
        visible.value = true
    }

    function close(){
        visible.value = false
    }

    return {
        visible,
        show,
        close
    }
})