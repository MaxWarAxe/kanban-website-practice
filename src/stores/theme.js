import {defineStore} from 'pinia'
import {ref} from 'vue'

export const useThemeStore = defineStore('theme', () => {
    let dark = ref(false)
    let light = ref(false)

    function init(){
        const state1 = localStorageGet()
        if(state1 != undefined){
            dark.value = state1.dark
            light.value = state1.light
            console.log('storageUseed')
        }
        else{
            let darkColorScheme = window.matchMedia('(prefers-color-scheme: dark)')
            dark.value = darkColorScheme.matches
            light.value = !darkColorScheme.matches
            localStorageSave()
        }
        applyClasses()
    }

    function toggleTheme(){
        dark.value = !dark.value
        light.value = !light.value
        applyClasses()
        localStorageSave()
    }

    function applyClasses(){
        document.body.classList.remove('light-theme', 'dark-theme');
        if (dark) {
            document.body.classList.add('dark-theme');
        } else if (light) {
            document.body.classList.add('light-theme');
        }
    }

    function localStorageSave(){
        const stateToSave = {dark : dark.value, 
                            light : light.value}
        localStorage.setItem('THEME_STATE',JSON.stringify(stateToSave));
    }

    function localStorageGet(){
        return JSON.parse(localStorage.getItem('THEME_STATE'))
    }
    
    return {
        dark,
        light,
        init,
        toggleTheme,
        applyClasses,
        localStorageSave,
        localStorageGet
    }
})