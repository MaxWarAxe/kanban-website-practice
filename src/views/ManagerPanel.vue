<script setup lang="ts">
    import { useTaskPanelStore } from '@/stores/taskPanel';
    import { RouterView } from 'vue-router';

    const taskPanelStore = useTaskPanelStore()
</script>

<template>
  <div class="task-panel">
    <div
      class="task-panel__cross-button"
      @click="taskPanelStore.close()"
      data-test="cross-button"
    >
      <svg
        class="task-panel__cross-icon"
        viewBox="0 0 24 24"
        width="32"
        height="32"
      >
        <use href="@/assets/icons/themeIcon.svg#cross-icon" />
      </svg>
    </div>
    <RouterView />
  </div>
</template>

<style lang="scss" scoped>
    .task-panel{
        width: 350px;
        height: 100vh;
        top: 0px;
        resize: horizontal;
        overflow: auto;
        right: 0px;
        position: absolute;
        background-color: var(--primary-container-color);
        border-left: solid;
        border-color: var(--primary-border-color);
    }
    .task-panel__cross-icon{
        fill: var(--primary-icon-color)
    }
</style>