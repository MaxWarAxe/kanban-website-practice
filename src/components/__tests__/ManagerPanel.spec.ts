import { describe,it,expect, vi, beforeEach } from 'vitest'
import { mount } from '@vue/test-utils'
import { createTestingPinia } from '@pinia/testing'
import { setActivePinia,createPinia } from 'pinia'
import { useTaskPanelStore } from '@/stores/taskPanel'
import ManagerPanel from '@/views/ManagerPanel.vue'
import { nextTick } from 'vue'
import router from '@/router/index.ts'

describe('ManagerPanel tests', () => {
    beforeEach(() => {
        setActivePinia(createPinia())
    })  

    it('Корректно рендерится', async () => {
        const wrapper = mount(ManagerPanel, {
            global: {
                plugins: [
                    createTestingPinia({createSpy: vi.fn, stubActions: false}),
                    router
                ]
            }
        })
        await nextTick()
        expect(wrapper.html()).toMatchSnapshot()
    })

    it('Закрывается по кнопке', async () => {
        const wrapper = mount(ManagerPanel, {
            global: {
                plugins: [
                    createTestingPinia({createSpy: vi.fn, stubActions: false}),
                    router
                ]
            }
        })
        const taskPanelStore = useTaskPanelStore()
        const button = wrapper.get('[data-test="cross-button"]')
        button.trigger('click')
        await nextTick()
        expect(taskPanelStore.visible).toBeFalsy()
    })
})  