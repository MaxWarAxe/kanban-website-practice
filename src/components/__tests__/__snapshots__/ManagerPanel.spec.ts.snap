// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ManagerPanel tests > Корректно рендерится 1`] = `
"<div data-v-39bc4ccd="" class="task-panel">
  <div data-v-39bc4ccd="" class="task-panel__cross-button" data-test="cross-button"><svg data-v-39bc4ccd="" class="task-panel__cross-icon" viewBox="0 0 24 24" width="32" height="32">
      <use data-v-39bc4ccd="" href="/src/assets/icons/themeIcon.svg#cross-icon"></use>
    </svg></div>
  <!---->
</div>"
`;
